File: 0101-0200/0133-Clone Graph/0133-Clone Graph.py
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = []):
        self.val = val
        self.neighbors = neighbors

import collections
class Solution:
    def cloneGraph(self, node: 'Node') -> 'Node':
        if node is None:
            return None
        table = {node: Node(node.val, [])}
        Q = collections.deque([node])
        while Q:
            n = Q.popleft()
            for neighbor in n.neighbors:
                if neighbor not in table:
                    table[neighbor] = Node(neighbor.val, [])
                    Q.append(neighbor)
                table[n].neighbors.append(table[neighbor])
        return table[node]

File: 0101-0200/0133-Clone Graph/0133-Clone Graph.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/

class Solution {
public:
    Node* cloneGraph(Node* node) {
        if (node == nullptr) {
            return nullptr;
        }
        unordered_map<Node*, Node*> table;
        table[node] = new Node(node->val);
        queue<Node*> Q;
        Q.push(node);
        while (!Q.empty()) {
            Node* n = Q.front();
            Q.pop();
            for (Node* neighbor : n->neighbors) {
                if (!table.count(neighbor)) {
                    table[neighbor] = new Node(neighbor->val);
                    Q.push(neighbor);
                }
                table[n]->neighbors.push_back(table[neighbor]);
            }
        }
        
        return table[node];
    }
};
